howso.amlg

```diff
@@ -90,18 +94,258 @@
 			(retrieve_from_entity trainee "pointVersion")
 		)
 
+	;helper method used to determine the full hierarchy path to a subtrainee used to create or load subtrainees
+	;outputs path in the format of: (list trainee ".trainee_container" child ".trainee_container" grandchild ...)
+	;outputs a string error if invalid parameters are provided
+	#!determine_child_trainee_path
+	(seq
+		(if (and (or child_name_path child_id) (not (contains_entity trainee)))
+			(conclude "Specified invalid trainee.")
+		)
+		(if (and (= (null) child_id) (= 0 (size child_name_path)))
+			(conclude (null))
+		)
+
+		(declare (assoc new_child_name (get payload "trainee") ))
+
+		(declare (assoc
+			path_to_parent
+				(if (size child_name_path)
+					(call_entity trainee "ResolveChildPath" (assoc name_path child_name_path))
+
+					(call_entity trainee "ResolveChildPath" (assoc id child_id))
+				)
+		))
+
+		(if (!~ (list) path_to_parent)
+			(conclude "Specified invalid child trainee.")
+		)
+
+		(if (contains_entity (append trainee path_to_parent ".trainee_container" new_child_name))
+			(conclude "Specified trainee is already in the hierarchy.")
+		)
+
+		(append path_to_parent ".trainee_container" new_child_name)
+	)
+
+	;Pull the hierarchy for a trainee, returns an assoc of:
+	;'hierarchy_map' which reperesents the hierarchy tree and 'is_contained_map' which is an assoc of subtrainee -> is_contained flag
+	;parameters:
+	; trainee: id of trainee whose hierarchy info to return
+	#get_hierarchy
+		(call !return (assoc
+			payload
+				(assoc
+					"hierarchy_map" (retrieve_from_entity trainee "hierarchyMap")
+					"is_contained_map" (retrieve_from_entity trainee "childTraineeIsContainedMap")
+				)
+		))
+
+	;Rename a contained trainee
+	;
+	;parameters:
+	; trainee: name of trainee whose contained trainee should be renamed
+	; child_name_path: optional, list of strings specifying the user-friendly path of the child subtrainee to rename
+	; child_id: optional, id of child trainee to rename. Ignored if child_name_path is specified
+	; new_name: new name of child trainee
+	#rename
+	(seq
```

Comment:
it's because all the parameters are defaulted to null for this method, so there's no need to explicitly declare them as such.  I'll go ahead and update it to the declare paradigm anyway.